source/extensions/filters/network/dubbo_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,352 @@
+    conn_pool_handle_ = handle;
+    return Network::FilterStatus::StopIteration;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void Router::UpstreamRequest::resetStream() {
+  if (conn_pool_handle_) {
+    ASSERT(!conn_data_);
+    conn_pool_handle_->cancel(Tcp::ConnectionPool::CancelPolicy::Default);
+    conn_pool_handle_ = nullptr;
+  }
+
+  if (conn_data_) {
+    ASSERT(!conn_pool_handle_);
+    conn_data_->connection().close(Network::ConnectionCloseType::NoFlush);
+    conn_data_.reset();
+  }
+}
+
+void Router::UpstreamRequest::encodeData(Buffer::Instance& data) {
+  if (!conn_data_) {
+    ENVOY_STREAM_LOG(trace, "buffering {} bytes", *parent_.callbacks_, data.length());
+    if (!buffered_request_body_) {
+      buffered_request_body_ = std::make_unique<Envoy::Buffer::OwnedImpl>();
+    }
+
+    buffered_request_body_->move(data);
+  } else {
+    ENVOY_STREAM_LOG(trace, "proxying {} bytes", *parent_.callbacks_, data.length());
+    conn_data_->connection().write(data, false);
+  }
+}
+
+void Router::UpstreamRequest::onPoolFailure(Tcp::ConnectionPool::PoolFailureReason reason,
+                                            Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(warn, "dubbo upstream request: connection failure '{}'", host->address()->asString());
+
+  conn_pool_handle_ = nullptr;
+
+  // Mimic an upstream reset.
+  onUpstreamHostSelected(host);
+  onResetStream(reason);
+
+  parent_.upstream_request_buffer_.drain(parent_.upstream_request_buffer_.length());
+
+  // If it is a connection error, it means that the connection pool returned
+  // the error asynchronously and the upper layer needs to be notified to continue decoding.
+  // If it is a non-connection error, it is returned synchronously from the connection pool
+  // and is still in the callback at the current Filter, nothing to do.
+  if (reason == Tcp::ConnectionPool::PoolFailureReason::Timeout ||
+      reason == Tcp::ConnectionPool::PoolFailureReason::LocalConnectionFailure ||
+      reason == Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure) {
+    parent_.callbacks_->continueDecoding();
+  }
+}
+
+void Router::UpstreamRequest::onPoolReady(Tcp::ConnectionPool::ConnectionDataPtr&& conn_data,
+                                          Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(debug, "dubbo upstream request: tcp connection has ready");
+
+  // Only invoke continueDecoding if we'd previously stopped the filter chain.
+  bool continue_decoding = conn_pool_handle_ != nullptr;
+
+  onUpstreamHostSelected(host);
+  conn_data_ = std::move(conn_data);
+  conn_data_->addUpstreamCallbacks(parent_);
+  conn_pool_handle_ = nullptr;
+
+  onRequestStart(continue_decoding);
+}
+
+void Router::UpstreamRequest::onRequestStart(bool continue_decoding) {
+  ENVOY_LOG(debug, "dubbo upstream request: start sending data to the server {}",
+            upstream_host_->address()->asString());
+
+  if (buffered_request_body_) {
```

Comment:
What is the purpose of buffered_request_body_?

Reply:
The goal is to cache the transfer data when the connection is not ready and the Filter's callback interface is complete.